identity-core/src/crypto/key/pair.rs

```diff
@@ -24,14 +23,14 @@ pub struct KeyPair {
 
 impl KeyPair {
   /// Creates a new [`Ed25519`][`KeyType::Ed25519`] [`KeyPair`].
-  pub fn new_ed25519() -> Result<Self> {
+  pub fn new_ed25519() -> Result<Self, Ed25519KeyPairGenerationError> {
     Self::new(KeyType::Ed25519)
   }
 
   /// Creates a new [`KeyPair`] with the given [`key type`][`KeyType`].
-  pub fn new(type_: KeyType) -> Result<Self> {
+  pub fn new(type_: KeyType) -> Result<Self, Ed25519KeyPairGenerationError> {
```

Comment:
This is not forwards compatible, if we introduce more `KeyType`s. The returned error should correspond to the respective `KeyType`. A non-exhaustive `KeyPairGenerationError` enum could be added that wraps `Ed25519KeyPairGenerationError` and others in the future. But then we have the enum problem again, which is not ideal here.

We could also think about making `KeyType` generic, and then we could return the associated error type. This would probably include a larger refactor of how we represent keys, too. For example, having a dedicated `Ed25519PrivateKey` that enforces the 32 byte length. That would also be useful to have in the account, so we could take a valid ed25519 key as the input to a method creation, without having to do internal validation ourselves. (This is where `try_from_ed25519_bytes` comes from, and more errors on the account level.) I haven't thought through all the ramifications of that, though.

Reply:
As I don't think people care much about why/how keypair generation failed (at least not at run time), I think I will just rename `Ed25519KeyPairGenerationError` to `KeyPairGenerationError`. What do you think of that solution?

Reply:
Good point, actually. I think failure is mostly I/O related, so a plain `KeyPairGenerationError` should work.

Since I have changed my opinion on associated error types anway, this seems like a good decision :smile:.